Geometry/HGCalCommonData/src/HGCalGeomParameters.cc

```diff
@@ -261,13 +261,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
+  // were) to ensure that new copy numbers are always added to the end of the
+  // list.
+  std::unordered_map<int32_t, int32_t> copies;
+  HGCalParameters::layer_map copiesInLayers(layers.size() + 1);
+  std::vector<int32_t> wafer2copy;
+  std::vector<HGCalGeomParameters::cellParameters> wafers;
+  cms::DDFilteredView fv1(cpv->detector(), cpv->detector()->worldVolume());
+  std::string attribute = "Volume";
+  cms::DDSpecParRefs ref1;
+  const cms::DDSpecParRegistry& mypar1 = cpv->specpars();
+  mypar1.filter(ref1, attribute, sdTag2);
+  fv1.mergedSpecifics(ref1);
+  bool ok = fv1.firstChild();
+  if (!ok) {
+    edm::LogError("HGCalGeom") << " Attribute " << sdTag2 << " not found but needed.";
+    throw cms::Exception("DDException") << "Attribute " << sdTag2 << " not found but needed.";
+  } else {
+    dodet = true;
+    std::unordered_set<std::string> names;
+    while (dodet) {
+      const std::string name = static_cast<std::string>(fv1.name());
+      std::vector<int> copy = fv1.history().copyNos;
+      int nsiz = (int)(copy.size());
+      int wafer = (nsiz > 0) ? copy[nsiz - 1] : 0;
+      int layer = (nsiz > 1) ? copy[nsiz - 2] : 0;
+      if (nsiz < 2) {
+        edm::LogError("HGCalGeom") << "Funny wafer # " << wafer << " in " << nsiz << " components";
+        throw cms::Exception("DDException") << "Funny wafer # " << wafer;
+      } else if (layer > (int)(layers.size())) {
+        edm::LogWarning("HGCalGeom") << "Funny wafer # " << wafer << " Layer " << layer << ":" << layers.size()
+                                     << " among " << nsiz << " components";
+      } else {
+        auto itr = copies.find(wafer);
+        auto cpy = copiesInLayers[layer].find(wafer);
+        if (itr != copies.end() && cpy == copiesInLayers[layer].end()) {
+          copiesInLayers[layer][wafer] = itr->second;
+        }
+        if (itr == copies.end()) {
+          copies[wafer] = wafer2copy.size();
+          copiesInLayers[layer][wafer] = wafer2copy.size();
+          double xx = HGCalParameters::k_ScaleFromDD4Hep * fv1.translation().X();
+          if (std::abs(xx) < tolerance)
+            xx = 0;
+          double yy = HGCalParameters::k_ScaleFromDD4Hep * fv1.translation().Y();
+          if (std::abs(yy) < tolerance)
+            yy = 0;
+          wafer2copy.emplace_back(wafer);
+          GlobalPoint p(xx, yy, HGCalParameters::k_ScaleFromDD4Hep * fv1.translation().Z());
+          HGCalGeomParameters::cellParameters cell(false, wafer, p);
+          wafers.emplace_back(cell);
+          if (names.count(name) == 0) {
+            double zv[2], rv;
+            const std::vector<double>& pars = fv1.parameters();
+            if (mode == HGCalGeometryMode::Polyhedra) {
+              zv[0] = pars[4];
+              zv[1] = pars[7];
+              rv = pars[6];
+            } else {
+              zv[0] = pars[4];
+              zv[1] = pars[7];
+              rv = pars[6];
```

Comment:
Please add FIXME comment here also.

Reply:
I shall remember fixing these once you have done for the other shapes